source/common/config/config_provider_impl.h

```diff
@@ -0,0 +1,410 @@
+namespace Config {
+
+// This file provides a set of base classes, (ImmutableConfigProviderImplBase,
+// MutableConfigProviderImplBase, ConfigProviderManagerImplBase, ConfigSubscriptionInstanceBase),
+// conforming to the ConfigProvider/ConfigProviderManager interfaces, which in tandem provide a
+// framework for implementing statically defined (i.e., immutable) and dynamic (mutable via
+// subscriptions) configuration for Envoy.
+//
+// Dynamic configuration is distributed via xDS APIs (see
+// https://github.com/envoyproxy/data-plane-api/blob/master/XDS_PROTOCOL.md). The framework exposed
+// by these classes simplifies creation of client xDS implementations following a shared ownership
+// model, where according to the config source specification, a config subscription, config protos
+// received over the subscription and the subsequent config "implementation" (i.e., data structures
+// and associated business logic) are shared across ConfigProvider objects and Envoy worker threads.
+//
+// This approach enables linear memory scalability based primarily on the size of the configuration
+// set.
+//
+// A blueprint to follow for implementing {im,}mutable config providers is as follows:
+//
+// For both:
+//   1) Create a class derived from ConfigProviderManagerImplBase and implement the required
+//   interface.
+//      When implementing createXdsConfigProvider(), it is expected that getSubscription<T>() will
+//      be called to fetch either an existing ConfigSubscriptionInstanceBase if the config source
+//      configuration matches, or a newly instantiated subscription otherwise.
+//
+// For immutable providers:
+//   1) Create a class derived from ImmutableConfigProviderImplBase and implement the required
+//   interface.
+//
+// For mutable (xDS) providers:
+//   1) Create a class derived from MutableConfigProviderImplBase and implement the required
+//   interface.
+//   2) Create a class derived from ConfigSubscriptionInstanceBase; this is the entity
+//   responsible for owning and managing the Envoy::Config::Subscription<ConfigProto> that provides
+//   the underlying config subscription.
+//     - When subscription callbacks (onConfigUpdate, onConfigUpdateFailed) are issued by the
+//     underlying subscription, the corresponding ConfigSubscriptionInstanceBase functions must be
+//     called as well.
+//     - On a successful config update, checkAndApplyConfig() should be called to instantiate the
+//     new config implementation and propagate it to the shared config providers and all
+//     worker threads.
+
+class ConfigProviderManagerImplBase;
+
+/**
+ * Specifies the type of config associated with a ConfigProvider.
+ */
+enum class ConfigProviderInstanceType {
+  // Configuration defined as a static resource in the bootstrap config.
+  Static,
+  // Configuration defined inline in a resource that may be specified statically or obtained via
+  // xDS.
+  Inline,
+  // Configuration obtained from an xDS subscription.
+  xDS
```

Comment:
Nit: prefer `Xds` for consistency here for the enum class value.

Reply:
Done.